streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -989,16 +989,20 @@ private StreamThread createAndAddStreamThread(final long cacheSizePerThread, fin
     public Optional<String> removeStreamThread() {
         if (isRunningOrRebalancing()) {
             synchronized (changeThreadCount) {
-                for (final StreamThread streamThread : threads) {
-                    if (streamThread.isAlive() && (!streamThread.getName().equals(Thread.currentThread().getName()) || threads.size() == 1)) {
-                        streamThread.shutdown();
-                        if (!streamThread.getName().equals(Thread.currentThread().getName())) {
-                            streamThread.waitOnThreadState(StreamThread.State.DEAD);
+                synchronized (threads) {
```

Comment:
Why are we using the threads object as a lock here? is that was the `Collections.synchronizedList(new LinkedList<>());` uses internally?

I am a little concerned that we are holding `changeThreadCount` while waiting for `threads`

Reply:
make sense. Let me use 'copy' to handle lock issue.